scanpipe/pipes/d2d.py

```diff
@@ -496,30 +493,107 @@ def create_package_from_purldb_data(project, resource, package_data):
     return package
 
 
-def match_purldb_package(project, resource):
-    """Match an archive type resource in the PurlDB."""
-    if results := purldb.match_package(sha1=resource.sha1):
-        package_data = results[0]
-        return create_package_from_purldb_data(project, resource, package_data)
+def process_purldb_package_data(project, package_data, resources):
+    """
+    Given a mapping of `package_data`, create a Package from `package_data`, and
+    associate the Package to Resources in `resources`.
+
+    Return the number of Resources matched to a Package.
+    """
+    match_resources_count = 0
+    for resource in resources:
+        matched_to_purldb_before = resource.status == flag.MATCHED_TO_PURLDB
+        matched_package = create_package_from_purldb_data(
+            project=project, resource=resource, package_data=package_data
+        )
+        resource.refresh_from_db()
+        matched_to_purldb_after = resource.status == flag.MATCHED_TO_PURLDB
+        if matched_package and matched_to_purldb_before != matched_to_purldb_after:
+            # Increment matched_resources_count if we have `matched_package` and
+            # when we are able to match on a Resource for the first time.
+            match_resources_count += 1
+    return match_resources_count
```

Comment:
it seems that the `pipes.update_or_create_package` will be called for each single resource. This is not very efficient. We should call the package creation code once, then deal with the resource association with an optimized queryset.